### DISCUSSION TITLE
トークナイザーとステミングの組み合わせについて

### DISCUSSION THREAD

**Discussion by takasegawa** [pgroonga/pgroonga#275] [Q&A] [answered]
お世話になっております。
今回はトークナイザーとステミングの組み合わせについて質問させてください。

現在、下記のインデックスを使用して半角英数記号の全文検索を実装しようとしています。
`create index idx_pgroonga_eigo on test.tabel1 using pgroonga ( eigo) with (plugins='token_filters/stem', tokenizer='TokenBigramSplitSymbolAlphaDigit', token_filters='TokenFilterStem');`

上記で想定している検索結果はほぼ得られているのですが、ステミング検索のみ実現できていない状況です。
そこで、下記について確認させていただけないでしょうか。

①ステミング検索ができない理由は、下記のリファレンスにある記載の通り、英単語が2文字単位でトークナイズされているため、ステミング対象として扱われていないことが原因で相違ないでしょうか。
>https://groonga.org/ja/docs/reference/token_filters/token_filter_stem.html
＞トークナイズされたトークンをステミングします。

②ステミング検索を実現するにあたり、どのような手法・設定ができるでしょうか。
こちらで検討した方法としては、下記のいずれかで対応できる認識です。

１．ステミング以外の全文検索をILIKE句で賄い、OR条件にステミング検索を追加することで実現
`select eigo from tabel1 where eigo ilike 'develops' OR eigo &@~ 'develops';`
　→ILIKEで全文検索相当の検索を賄い、ステミングの差分をPGroongaで補填
　→トークナイザーは「TokenBigramSplitSymbol」を使用

２．ステミングが効くインデックスと全文検索用のインデックスの2種類を作成し、
`create index idx_pgroonga_eigo1 on test.tabel1 using pgroonga ( eigo) with (plugins='token_filters/stem', tokenizer='TokenBigramSplitSymbol', token_filters='TokenFilterStem');`
`create index idx_pgroonga_eigo2 on test.tabel1 using pgroonga ( eigo) with (tokenizer='TokenBigramSplitSymbolAlphaDigit');`

上記以外で何か方法があればご教示いただけますと幸いです。

**Comment by komainu8** [comment, marked as answer]
①はご理解の通りです。
②は他にいい感じの方法がないか調べて見るので少々お待ちください。

**Reply by takasegawa** [reply]
ありがとうございます！何卒よろしくお願い致します。

**Reply by komainu8** [reply]
検討してみました。2の複数インデックスを用意するのが良さそうに思います。

通常ステミングは、対象が文章や文ではなく、単語単位なのでトークナイズ後に実行されます。
アルファベットを2文字単位で分割する場合はステミングは効果を発揮できません。
なので、`TokenFilterStem`と`TokenBigramSplitSymbolAlphaDigit`を同時に使って
どちらの効果も得ようとするのは1つのインデックスでは実現できません。

1の方法はILIKEもPGroongaのインデックスを使うので期待する結果は取得できないと思います。
（ちなみになのですが、PGroongaはデフォルトで大文字、小文字を区別せず検索するのでLIKEでも同じ結果がとれると思います。）
2のように2種類のインデックスを作って、検索時に使うインデックスを切り替えるようにするなら、実現できます。

やり方ですが、ご提案の方法ですと、適切なインデックスを使用してくれないので以下のような形にします。

1. ステミングを使って検索するためのインデックスを作成

```sql
CREATE INDEX idx_pgroonga_stem ON test.tabel1 USING pgroonga ((eigo || '')) WITH (plugins='token_filters/stem', tokenizer='TokenBigramSplitSymbol', token_filters='TokenFilterStem')
```

2. `TokenBigramSplitSymbolAlphaDigit`を使って検索するためのインデックスを作成

```sql
CREATE INDEX idx_pgroonga_bigram ON test.tabel1 USING pgroonga ((eigo || '' || '')) WITH (tokenizer='TokenBigramSplitSymbolAlphaDigit');
```

3. 1,2のインデックスを使って検索

```sql
SELECT eigo FROM tabel1 WHERE (eigo || '') &@~ 'develops' OR (eigo || '' ||'') &@~ '|agr';
 id |           content           
----+-----------------------------
  1 | I develop Groonga.
  2 | I developped Groonga.
  3 | I develops Groonga.
  4 | |agreement|
  5 | |AGR type reactor|
  6 | |agri-environmental policy|
(6 rows)
```

PostgreSQLは、字面でどのインデックスが使用候補になるのかを判断しています。
式インデックスというものを使って同じ値ではあるが、異なる字面のインデックスを作ることでどのインデックスを使うかをユーザーが制御できます。

したがって、`(eigo || '')`と`(eigo || '' || '')`を作ります。
`|| ''` と `|| '' || ''` に意味はありません。字面を変えるためだけに記載しています。

ステミングを使って検索したい場合は、`(column || '') &@~ ...`を`TokenBigramSplitSymbolAlphaDigit`を使いたい場合は、`(column || '' || '') &@~ ...`を使います。

**Reply by komainu8** [reply]
> （ちなみになのですが、PGroongaはデフォルトで大文字、小文字を区別せず検索するのでLIKEでも同じ結果がとれると思います。）

↑ですが、補足します。
LIKE/ILIKEを使った時は、大文字小文字が違う以外のPGroongaで指定したノーマライザーのおかげでヒットしたレコードは除外されます。

**Reply by takasegawa** [reply]
@komainu8 様
手法のご提案に加え詳細な解説まで頂き、誠にありがとうございました。
早速ご提案頂いた方法で動作確認を行ってみますので、申し訳ありませんが結果が出るまで今しばらくお待ちください。
（本番環境の都合で来週の火曜まで検証ができず…）

**Reply by komainu8** [reply]
承知しました！

**Reply by askdkc** [reply]
@komainu8 さん

割込失礼します
m(_ _)m

> ```sql
> 
> CREATE INDEX idx_pgroonga_bigram ON test.tabel1 USING pgroonga ((eigo || '' || '')) WITH (tokenizer='TokenBigramSplitSymbolAlphaDigit');
> 
> ```

このSQLを見て思ったのですがPGroongaってテーブルのカラム名ではなく、カラム用に作成したインデックスをwhere句に指定して検索出来ちゃうってことですか？

(間違ってること聞いてたらすみません)

**Reply by komainu8** [reply]
ちょっと説明しにくいので、説明がわかりにくかったらすみませんん。

インデックスをWHERE句に設定するというか、WHERE句の式に対してインデックスを設定しています。
これは、PGroongaではなくPostgreSQLの機能でして、式インデックスというものです。
これは、あるデータに対して関数などを適用した結果にインデックスを適用するというものです。

WHERE column = 'xxx' みたいな条件のときは、columnにインデックスを設定しますが、例えばcolumnの値を小文字にして検索したい時、にWHERE lower(column) = 'xxx' のようにPostgreSQLのlower()という関数を使って検索するとします。(PGroonga使えば小文字で検索できますが、説明のため、PGroongaは使わないということにしてください。。。)このときは、lower(column)の結果に対してインデックスを設定します。(columnにインデックスを設定しても、lower()関数によって、columnの値が変わるためcolumnに設定したインデックスが効きません。 )
で、この時lower(column)の結果に対してインデックスを設定して、インデックスを使った検索をできるようにするために式インデックスを使います。例えば、usersテーブルにnameというカラムがあったとして、`lower(column)`の結果にインデックスを設定したいときは以下のように書きます。

```
CREATE INDEX idx_lower_name ON users ((lower(name)));
```

式インデックスなので、インデックスを設定したい式を指定します。上記の例なら`lower(column)`です。
インデックスの設定対象は式なので、WHERE句に書いた式（関数）をインデックスの対象としている。という表現でわかっていただけるでしょうか。。。（うまく説明できている気がしませんが）このような説明でご質問に答えられているでしょうか。。。？

**Reply by askdkc** [reply]
ありがとうございます🙏

なんとなく理解できた、、、はず。

週末に何度か読み返して理解に咀嚼出来るように努めたいと思います🫠

**Reply by kou** [reply]
式インデックスの公式ドキュメント：

* https://www.postgresql.org/docs/current/indexes-expressional.html
* https://www.postgresql.jp/document/14/html/indexes-expressional.html

**Comment by takasegawa** [comment]
@komainu8 
前回の回答から日が経ってしまい申し訳ありません。
結合環境・本番環境にて上記方法を適用したところ、想定通りの検索結果を得ることができました。
ご回答いただきまして誠にありがとうございました。

しかし、検証中に1点原因不明の事象が発生いたしました。
結合環境ではスキーマ「test」を指定しなければ作成したPGroongaインデックスを認識しなかったため、
下記のようにスキーマ指定をSQLに埋め込んでいました。
`【例１】
SET search_path to test;
SELECT eigo FROM test.tabel1 WHERE (eigo || '') &@~ 'develops' OR (eigo || '' ||'') &@~ '|agr';`

上記のソースを本番環境にデプロイして実行したところ、演算子「&@~」が認識できない旨のエラーが発生しました。
リリース前のソースとの差異を考慮して検証したところ、下記のように「public」スキーマを指定すれば
動作することが確認できました。

`【例２】
SET search_path to test, public;
SELECT eigo FROM test.tabel1 WHERE (eigo || '') &@~ 'develops' OR (eigo || '' ||'') &@~ '|agr';`

publicスキーマを指定しなければ演算子を認識できない本番環境と、publicスキーマを指定しなくても
演算子を認識できる結合環境とで、何か考えられる違いはありますでしょうか。
直接的な回答でなくても、何か手掛かりとなる情報だけでも頂ければ幸いです。
確認に必要な情報があればこちらで収集させていただきますので、何卒ご協力をお願い致します。

※バージョンに差異はないことは確認しております。

**Reply by kou** [reply]
PGroongaの演算子は`CREATE EXTENSION pgroonga`したときのスキーマに定義されます。

結合環境では現在のスキーマが`test`の状態で`CREATE EXTENSION pgroonga`していて本番環境では現在のスキーマが`public`の状態で`CREATE EXTENSION pgroonga`しているんだと思います。

**Reply by takasegawa** [reply]
@kou 
早急なご回答誠にありがとうございました。
早速`CREATE EXTENSION pgroonga`を実行するスキーマをpublicに変更したところ、`SET search_path`を指定せずともインデックスを認識するようになりました。

本件解消したことで、環境差異として保留していた事項が同時に解消いたしました。
ご協力いただいき誠にありがとうございました！

**Reply by kou** [reply]
それはよかったです。

https://pgroonga.github.io/ja/users/ で事例を集めているので運用が開始されたらぜひご協力ください。サービス名を伏せた上で、どのように使っているか・どのあたりがよかったかなどを紹介してもらえるのでも大丈夫です。

事例紹介を収集している目的は新規ユーザー獲得促進です。
